Prettify the UI for Search Results

andres-holguin opened the pull request:
Note: Since this PR pulled from the currently ongoing PRs, it seems overloaded with commits. 
# UI Improvements

![Starfish-Slide-Group-V1](https://user-images.githubusercontent.com/44207690/88546324-35cac700-cfea-11ea-8e54-b44b02ca2b14.gif)

#### EDIT: Added card highlighting
![Starfish-Slide-Group-V2](https://user-images.githubusercontent.com/44207690/88815997-ef13d300-d189-11ea-9bdd-5fc38779635b.gif)

This PR brings some improvements to the UI for how Note results are displayed:

### Main features
- NoteGridCollections can be navigated as a slide group with arrows on both sides.
- Long NoteGridCard titles and school names will be truncates with ellipses (...) at the end
- The thumbnail image is now contained in a <v-img> component, which allows for easy overlays
- The icons used to represent google drive files and pdf files have been changed to use material design icons.
- EDIT: Cards will be highlighted/activated when clicked on. Only activated cards open the preview on click.

### TODO:
- Set Date parsing so certain time frames are displayed as "x days/hours ago"
- Decide how to best indicate NoteRating on NoteGridCard
- EDIT: Compute whether a note is a GNote

keelywan committed:
Merge remote-tracking branch 'origin/integration-fixes' into fav-download-logic

keelywan committed:
Create FavoriteNoteServlet

keelywan committed:
Merge remote-tracking branch 'origin/upload-notes' into fav-download-logic

keelywan committed:
Add fetch url to favorite note servlet

keelywan committed:
Fix SQL query in delete method

keelywan committed:
Add download button to note preview

keelywan committed:
Create DownloadNoteServlet

keelywan committed:
Fix comment for favorite note servlet

keelywan committed:
Add increment downloads function in NoteService

keelywan committed:
Add DataSource import to download note servlet

keelywan committed:
Connect NotePreview with DownloadNoteServlet

keelywan committed:
Add getAuthorId method in Note model

keelywan committed:
Add getNoteByNoteId method in NoteService

keelywan committed:
Get the userId of the note that was favorited/downloaded in servlet

keelywan committed:
Rename userId variable and add NoteService to FavoriteNoteServlet

keelywan committed:
Merge remote-tracking branch 'origin/upload-notes' into fav-download-logic

keelywan committed:
Add break to cookie search

andres-holguin committed:
Merge branch 'backend-integration'

andres-holguin committed:
Create test.html with routing example

Aradhya-B committed:
Add date column to favorite_notes table

Aradhya-B committed:
Add date field to insert favorite note method that always inserts the current date

Aradhya-B committed:
Remove unused import

andres-holguin committed:
Add some vue grid components.

Add v-container to NoteGrid.vue, v-row to NoteGridCollection.vue, and v-card to NoteGridCard.vue

andres-holguin committed:
Resize GDrive icon

520px -> 18px height/width

Aradhya-B committed:
Add methods to favorite note service to return trending notes by timespan

Aradhya-B committed:
Add trending notes servlet that returns the most trending notes in a given timespan

Aradhya-B committed:
Add comments to trending notes methods

Aradhya-B committed:
Remove extra whitespace

Aradhya-B committed:
Remove unused println statement from favorite note service

andres-holguin committed:
Add pdf icon

andres-holguin committed:
Wrap NoteGridCard.vue in v-col

Wrap NoteGridCard in v-col

andres-holguin committed:
Merge branch 'navbar-fix' into prettify

andres-holguin committed:
Merge remote-tracking branch 'origin/fav-download-logic' into prettify

andres-holguin committed:
Merge remote-tracking branch 'upstream/trending-notes' into prettify

andres-holguin committed:
Merge remote-tracking branch 'upstream/favorite-logic' into prettify

andres-holguin committed:
Update MockSearchButton.vue

Increase the size of the mock data.

andres-holguin committed:
MockSearchButton: Add description comment

andres-holguin committed:
NoteGridCollection: Wrap in v-slide-group

andres-holguin committed:
NoteGridCollection: Always show slide arrows

andres-holguin committed:
NoteGridCard: Set v-card width

andres-holguin committed:
NoteGridCollection: Center on selected card

This also fixes a bug causing the note-grid-collection to appear empty when filters or search query was changes, when in reality the notes were hiding in the previous pane.

andres-holguin committed:
MockSearchButton: Enhance with long titles/schools

andres-holguin committed:
NoteGridCard: Replace div thumbnail with v-img

andres-holguin committed:
NoteGridCard: Move file icon into thumbnail

The icon that indicates whether a submission is a PDF or a Google Drive link is now overlayed in the top left corner of the thumbnail.

andres-holguin committed:
NoteGridCard: Vuetify the note descriptors

The title, course, and school are now wrapped in vuetify components that will leave ellipses if the content overflows.
The favorited icon now changes color based on props.

andres-holguin committed:
Merge remote-tracking branch 'upstream/master' into prettify

andres-holguin committed:
Delete test.html

Used for initial routing tests.

andres-holguin committed:
Remove unused icons in assets

andres-holguin committed:
Remove some file changes for pending commits.

keelywan commented on src/main/webapp/components/NoteGridCard.vue:
```diff
@@ -1,36 +1,63 @@
+    data: function() {
+      return {
+        favorited: this.isFavorited,
+        isGNote: true,
```
Do you think that this should be a computed value instead of in data?

andres-holguin replied:
Makes, sense did that and added a TODO for computing it :)

keelywan commented on src/main/webapp/components/NoteGridCard.vue:
```diff
@@ -1,36 +1,63 @@
+          <v-col col="1">
+            <v-badge overlap class="ma-1">
+              <template v-slot:badge>{{numFavorites}}</template>
+              <v-icon :color="favorited && 'yellow'">mdi-star</v-icon>
```
I tried running this locally, and I got an error saying that the type check failed for prop color because it expected a String but got a Boolean. Maybe you could store a color variable in data and then change it whenever `this.favorited` changes?

andres-holguin replied:
Ahh good catch, thought I could do something fancy but failed x) made that change.

keelywan commented on src/main/webapp/components/NoteGridCard.vue:
```diff
@@ -42,12 +69,20 @@
+        this.favorited = !this.favorited;
+        fetch('/favorite-note?note_id=' + this.id, {
+          method: method
+        })
```
I think this toggleFavorite method can be deleted because the toggling function should be in NotePreview instead.

andres-holguin replied:
Good catch

keelywan reviewed (approved):
The UI looks really good!

andres-holguin committed:
NoteGridCard.vue: Address Review

- Moved isGNote from data to computed, with TODO for actually checking if Note Data is for a GNote
- The color of the favorite star is now determined by a computed value, favColor
- Removed unused toggleFavorite method

andres-holguin committed:
NoteGridCard.vue: Remove unused styling

andres-holguin committed:
Highlight Active Cards

The active slot prop from v-slide-item is now used to highlight the active card slightly blue.

andres-holguin committed:
NoteGridCollection.vue: Fix highlight issue

Fixed issue where first card in collection was always highlighted when first rendered or the noteData changed. Now, the first card will be unselected at the next tick.

andres-holguin committed:
Modify how previews are opened.

Only active cards will display a preview on click. If not active, clicking on a card will make it active, first.

andres-holguin commented:
@keelywan re-requesting a review for new changes to card highlighting.

Aradhya-B reviewed (approved):
LGTM, but do we still need the MockSearchButton? I think now that normal search and note posting is working, we could probably do without it.

cftorres commented on src/main/webapp/components/MockSearchButton.vue:
```diff
@@ -46,6 +49,97 @@ module.exports = {
+          thumbnailSrc: "/assets/notes.png",
+          author: "Andres H.",
+          title: "ABC set of notes",
+          dateCreated: new Date(Date.now() - 1000 * 60 * 60 * 24 * 6).toDateString(),
```
prefer extracting this kind of math out into a constant (e.g. `SIX_DAYS` or something like `HOURS_PER_DAY * 6` if you'll be using this same constant in different ways) so it's clear how Date is being manipulated

andres-holguin replied:
You got it!

keelywan reviewed (approved):
Looks good! Thanks for making the changes 😄

cftorres reviewed (approved):
Looking good!

andres-holguin commented:
> LGTM, but do we still need the MockSearchButton? I think now that normal search and note posting is working, we could probably do without it.

I tend to run a web-server that isn't connected to the back end since it's faster for testing front-end changes.

andres-holguin committed:
MockSearchButton.vue: Use constants for lengths of time
